### PR TITLE
Python3 docs consolidation

### PR DESCRIPTION
**Pull request by loechel**
It should be just one documentation and that should be English. I moved all docs and translated them into English as good I could.

### CODE REVIEW

File: docs/RestrictedPython4/index.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+RestrictedPython 4.0.0 and upwords is a complete rewrite for Python 3 compatibility.
+
+As the ``compiler`` module has been declared deprecated in Python 2.6 and already been removed in Python 3.0 it has never been avaliable for any Python 3 version.
+
+
+Goals for a rewrite
+-------------------
+
+RestrictedPython is a core dependency for the Zope2 application server and therefore implicite for the content management system Plone.
+The Zope & Plone community want to continue there projects and as Python 2 is announced end of life by 2020, and will be ultimative replaced by Python 3 or higher.
```
Comment (jcbrand) [inline docs/RestrictedPython4/index.rst]: s/implicite/implicit
s/there projects/their projects
s/ultimative/ultimately

File: docs/RestrictedPython4/index.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+Zope and Plone should become Python 3 compatible.
+
+One of the core functionallity of Zope 2 and therefore Plone is the possibility to implement and modify Python Scripts and Templates through the web (TTW) without harming the application or server itself.
+
```
Comment (jcbrand) [inline docs/RestrictedPython4/index.rst]: s/functionallity/features
s/Scripts and Templates/scripts and templates

File: docs/RestrictedPython4/index.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+* 2.7
+* 3.4
+* 3.5
+* 3.6
+* PyPy2.7
+
+.. _`security support` : https://docs.python.org/devguide/index.html#branchstatus
+
+We do explicite exclude Python 3.3 and PyPy3 (which is based on the Python 3.3 specification) at the moment, because the changes from 3.4 upwords are huge and Python 3.3 is end of life by September 2017.
+
```
Comment (jcbrand) [inline docs/RestrictedPython4/index.rst]: s/do explicite/explicitly

File: .gitignore [hunk 4]
```diff
@@ -27,7 +27,7 @@ pyvenv.cfg
 /src/*.egg-info
 /var
 coverage.xml
-docs/Makefile
+#docs/Makefile
 docs/doctrees
 docs/html
-docs/make.bat
+#docs/make.bat
```
Comment (icemac) [inline .gitignore]: Please remove these lines which begin with a `#`.

File: docs/RestrictedPython4/index.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
+RestrictedPython 4+
+===================
+
+RestrictedPython 4.0.0 and upwords is a complete rewrite for Python 3 compatibility.
```
Comment (icemac) [inline docs/RestrictedPython4/index.rst]: s/upwords/upwards

File: docs/RestrictedPython4/index.rst [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+* 2.7
+* 3.4
+* 3.5
+* 3.6
+* PyPy2.7
+
+.. _`security support` : https://docs.python.org/devguide/index.html#branchstatus
+
+We do explicite exclude Python 3.3 and PyPy3 (which is based on the Python 3.3 specification) at the moment, because the changes from 3.4 upwords are huge and Python 3.3 is end of life by September 2017.
```
Comment (icemac) [inline docs/RestrictedPython4/index.rst]: s/do explicite/explicitly
PyPy3 is only currently based on Python 3.3.

File: docs/RestrictedPython4/index.rst [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+The following packages / modules have hard dependencies on RestrictedPython:
+
+* AccessControl -->
+* zope.untrustedpython --> SelectCompiler
+* DocumentTemplate -->
+* Products.PageTemplates -->
+* Products.PythonScripts -->
+* Products.PluginIndexes -->
+* five.pt (wrapping some functions and procetion for Chameleon) -->
```
Comment (icemac) [inline docs/RestrictedPython4/index.rst]: s/procetion/protection

File: docs/idea.rst [hunk 8]
```diff
@@ -0,0 +1,86 @@
+The Idea behind RestrictedPython
+================================
+
+Python is a `Turing-complete`_ programming language.
+To offer a Python interface for users in web context is a potential security risk.
+Web frameworks and Content Management Systems (CMS) want to offer their users a maximum extendability and that if possible through the web (TTW).
+This also means to have permissions to add functionallity via a Python Script.
+
+From the point of IT Secruity and safty there should be additional preventive measures taken to ensure integrity of the application and the server itself.
```
Comment (icemac) [inline docs/idea.rst]: s/Secruity/Security
s/safty/safety

File: docs/idea.rst [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+RestrictedPython did chose a way to define a safe subset of the python programming laguage.
+This is a common approach for securing a programming language.
+The `Ada Ravenscar profil`_ is a well known example of such an approach.
+
+Defining a secure subset of the language did work, by restricting the `EBNF`_ elements and explicite allow or disallow language features.
```
Comment (icemac) [inline docs/idea.rst]: s/explicite/explicitly

File: docs/idea.rst [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+As the power of a programming language came from its core library and implemented third party libraries, the call of those modules and methods must have an additional check and restricting.
```
Comment (icemac) [inline docs/idea.rst]: s/restricting/restriction

File: docs/RestrictedPython4/index.rst [hunk 11]
```diff
@@ -9,12 +9,11 @@ As the ``compiler`` module has been declared deprecated in Python 2.6 and alread
 Goals for a rewrite
 -------------------
 
-RestrictedPython is a core dependency for the Zope2 application server and therefore implicite for the content management system Plone.
-The Zope & Plone community want to continue there projects and as Python 2 is announced end of life by 2020, and will be ultimative replaced by Python 3 or higher.
+RestrictedPython is a core dependency for the Zope2 application server and therefore implicit for the content management system Plone.
+The Zope & Plone community want to continue their projects and as Python 2 is announced end of life by 2020, and will be ultimately replaced by Python 3 or higher.
 Zope and Plone should become Python 3 compatible.
```
Comment (jcbrand) [inline docs/RestrictedPython4/index.rst]: Sorry, this should be "ultimately be".

File: docs/idea.rst [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+As Python is a scripting language that is executed by an interpreter.
+Any Python code that should be executed have to be explict checked before executing a generated byte code by the interpreter.
+
+Python itself offers three methods that provide such a workflow:
+
+* ``compile()`` which compiles source code to byte code
+* ``exec`` / ``exec()`` which executes the byte code in the interpreter
+* ``eval`` / ``eval()`` which executes a byte code expression
+
+Therefore RestrictedPython offers a repacement for the python builtin function ``compile()`` (Python 2: https://docs.python.org/2/library/functions.html#compile / Python 3 https://docs.python.org/3/library/functions.html#compile).
```
Comment (icemac) [inline docs/idea.rst]: Please use links in ReST style to these compile functions.

File: docs/idea.rst [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+This method is defined as following:
+
+.. code:: Python
+
+    compile(source, filename, mode [, flags [, dont_inherit]])
+
+The definition of the ``compile()`` method has changed over time, but its relevant parameters ``source`` and ``mode`` still remains.
```
Comment (icemac) [inline docs/idea.rst]: s/remains/remain

File: docs/idea.rst [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+* ``'single'``
+
+For RestricedPython this ``compile()`` method is replaced by:
+
+.. code:: Python
+
+    compile_restriced(source, filename, mode [, flags [, dont_inherit]])
+
+The primary parameter ``source`` has to be a ASCII or ``unicode`` string.
+Both methods either returns compiled byte code that the interpreter could execute or throws errors if the provided source code is invalid.
```
Comment (icemac) [inline docs/idea.rst]: s/throws/throw

File: docs/idea.rst [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+As the ``compile`` and ``compile_restricted`` just compile the provided source code it is not enough to secure the system, as still all calls on the core library or third party libraries are available.
```
Comment (icemac) [inline docs/idea.rst]: s/As the/As

File: docs/idea.rst [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+* ``eval`` / ``eval()``
+
+did have two parameters:
+
+* globals
+* locals
+
+which are a reference to the Python builtins.
+
+By modifing and restricting the avaliable moules, methods and constans from globals and locals we could limit the possible calls.
```
Comment (icemac) [inline docs/idea.rst]: s/constans/constants

File: docs/idea.rst [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+* ``setattr``
+* ``import``
+
+Also RestrictedPython provides three predifined, limited versions of Python's own ``__builtins__``:
+
+* ``safe_builtins`` (by Guards.py)
+* ``limited_builtins`` (by Limits.py), which provides restriced sequence types
+* ``utilities_builtins`` (by Utilities.py), which provides access for standard modules math, random, string and for sets.
+
+Additional there exists guard functions to make attributes of Python objects immutable --> ``full_write_guard`` (write and delete protected)
```
Comment (icemac) [inline docs/idea.rst]: s/exists/exist

File: docs/RestrictedPython4/index.rst (hunk 11 above)
Comment (loechel) [inline docs/RestrictedPython4/index.rst]: @jcbrand any help is welcome. You could also do the correction yourself immediatly and push to this branch; additionally we are in te #sprint channel

File: docs/update/index.rst [hunk 18]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
 Produced byte code has to compatible with the execution environment, the Python Interpreter within this code is called.
 So we must not generate the byte code that has to be returned from ``compile_restricted`` and the other ``compile_restricted_*`` methods manually, as this might harm the interpreter.
```
Comment (icemac) [inline docs/update/index.rst]: s/must not/do not have to

File: docs/update/index.rst [hunk 19]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
 We actually don't even need that.
-The Python ``compile()`` function introduced the capability to compile ``ast.AST`` code into byte code.
+The Python ``compile()`` function introduced the capability to compile ``ast.AST`` code into byte code in Python 2.6.
+
+We do need to move from ``compiler.ast`` to ``ast.AST``.
```
Comment (icemac) [inline docs/update/index.rst]: s/ast.AST/ast

File: docs/update/index.rst [hunk 20]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
 
 
 compiler.ast --> ast.AST
```
Comment (icemac) [inline docs/update/index.rst]: s/ast.AST/ast

File: docs/update/index.rst [hunk 21]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
 ------------------------
 
-Aus Sicht des Compilerbaus sind die Konzepte der Module compiler und ast vergleichbar, bzw. ähnlich.
-Primär hat sich mit der Entwicklung von Python ein Styleguide gebildet wie bestimmte Sachen ausgezeichnet werden sollen.
-Während compiler eine alte CamelCase Syntax (``visitNode(self, node, walker)``) nutzt ist in AST die Python übliche ``visit_Node(self, node)`` Syntax heute üblich.
-Auch habe sich die Namen genändert, wärend compiler von ``Walker`` und ``Mutator`` redet heissen die im AST kontext ``NodeVisitor`` und ``NodeTransformator``
+From the point of view of compiler design the concepts of the compiler module and the ast module are similar.
+The ``compiler`` module seems to be a very old module of the Python history.
+There has happen several improvements of the Python development like a generall applied style guide.
```
Comment (icemac) [inline docs/update/index.rst]: s/has happen/have happened
s/generall/generally

File: docs/update/index.rst [hunk 22]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
+While ``compiler`` still uses the old `CamelCase`_ Syntax (``visitNode(self, node, walker)``) the ``ast.AST`` did now use the Python common ``visit_Node(self, node)`` syntax.
+Also the names has been changed, while ``compiler`` did talk about ``Walker`` and ``Mutator``the corrosponding elements in ``ast.AST`` are ``NodeVisitor`` and ``NodeTransformator``.
```
Comment (icemac) [inline docs/update/index.rst]: s/has/have
s/corrosponding/corresponding

File: docs/update/index.rst [hunk 23]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
 
 
 ast Modul (Abstract Syntax Trees)
 ---------------------------------
 
-Das ast Modul besteht aus drei(/vier) Bereichen:
+The ast module consist of three / four areas:
```
Comment (icemac) [inline docs/update/index.rst]: s/consist/consists
s/three \\/ four/four

File: docs/update/index.rst [hunk 24]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
 
-* AST (Basis aller Nodes) + aller Node Classen Implementierungen
-* NodeVisitor und NodeTransformer (Tools zu Ver-/Bearbeiten des AST)
-* Helper-Methoden
+* AST (Basis of all Nodes) + all node class implementation
```
Comment (icemac) [inline docs/update/index.rst]: s/implementation/implementations

File: docs/update/index.rst [hunk 25]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
   * PyCF_ONLY_AST
 
 
 NodeVisitor & NodeTransformer
 -----------------------------
 
-Ein NodeVisitor ist eine Klasse eines Node / AST Konsumenten beim durchlaufen des AST-Baums.
-Ein Visitor liest Daten aus dem Baum verändert aber nichts am Baum, ein Transformer - vom Visitor abgeleitet - erlaubt modifikationen am Baum, bzw. den einzelnen Knoten.
-
+A NodeVisitor is a class of a node / AST consumer, it just reads the data by steping through the tree but did not modify the tree.
```
Comment (icemac) [inline docs/update/index.rst]: s/did/does

File: docs/update/index.rst [hunk 26]
```diff
@@ -16,42 +14,44 @@ This machine code then gets executed on the specific CPU architecture, with all
...
+In oposite a NodeTransformer which inherit from a NodeVisitor is allowed to modify the tree and nodes.
```
Comment (icemac) [inline docs/update/index.rst]: s/inherit/inherits

**Pull request merged**